modules/substrate/src/lib.rs

```diff
@@ -14,5 +14,254 @@
 
+//! Substrate Bridge Pallet
+//!
+//! This pallet is an on-chain light client for chains which have a notion of finality.
+//!
+//! It has a simple interface for achieving this. First it can import headers to the runtime
+//! storage. During this it will check the validity of the headers and ensure they don't conflict
+//! with any existing headers (e.g they're on a different finalized chain). Secondly it can finalize
+//! an already imported header (and its ancestors) given a valid Grandpa justification.
+//!
+//! With these two functions the pallet is able to form a "source of truth" for what headers have
+//! been finalized on a given Substrate chain. This can be a useful source of info for other
+//! higher-level applications.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+// Runtime-generated enums
+#![allow(clippy::large_enum_variant)]
+
+use bp_substrate::{AuthoritySet, ImportedHeader, ScheduledChange};
+use frame_support::{decl_error, decl_module, decl_storage, dispatch};
+use frame_system::ensure_signed;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::{marker::PhantomData, prelude::*};
+
 mod justification;
 mod storage_proof;
+mod verifier;
+
+#[cfg(test)]
+mod mock;
+
+type Hash<T> = <T as HeaderT>::Hash;
+type Number<T> = <T as HeaderT>::Number;
+
+pub trait Trait: frame_system::Trait {}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as SubstrateBridge {
+		/// Hash of the best finalized header.
+		BestFinalized: T::Hash;
+		/// Headers which have been imported into the pallet.
+		ImportedHeaders: map hasher(identity) T::Hash => Option<ImportedHeader<T::Header>>;
+		/// The current Grandpa Authority set.
+		CurrentAuthoritySet: AuthoritySet;
+		/// The next scheduled authority set change.
+		///
+		// Grandpa doesn't require there to always be a pending change. In fact, most of the time
+		// there will be no pending change available.
+		NextScheduledChange: Option<ScheduledChange<Number<T::Header>>>;
+	}
+	add_extra_genesis {
+		config(initial_header): Option<T::Header>;
+		config(initial_authority_list): sp_finality_grandpa::AuthorityList;
+		config(initial_set_id): sp_finality_grandpa::SetId;
+		config(first_scheduled_change): Option<ScheduledChange<Number<T::Header>>>;
+		build(|config| {
+			assert!(
+				!config.initial_authority_list.is_empty(),
+				"An initial authority list is needed."
+			);
+
+			let initial_header = config
+				.initial_header
+				.clone()
+				.expect("An initial header is needed");
+
+			<BestFinalized<T>>::put(initial_header.hash());
+			<ImportedHeaders<T>>::insert(
+				initial_header.hash(),
+				ImportedHeader {
+					header: initial_header,
+					requires_justification: false,
+					is_finalized: true,
+				},
+			);
+
+			let authority_set =
+				AuthoritySet::new(config.initial_authority_list.clone(), config.initial_set_id);
+			CurrentAuthoritySet::put(authority_set);
+
+			let change = config.first_scheduled_change.clone();
+			if change.is_some() {
```

Comment:
By using `if let Some(ref change) = config.first_scheduled_change { ... }` you could remove that expect :)